### PR TITLE
support delta in pytorch training and get_raw_egs by chain2

### PR DESCRIPTION
**Pull request by fanlu**
- lda layer make the result a litter better 

### Result
***
||tdnn_1c_rd_rmc_rng_without_cleanup | tdnn_1c_rd_rmc_rng|
|--|-- | --|
|dev_cer | 5.92 | 5.99|
|dev_wer | 13.71 | 13.86|
|test_cer | 7.03 | 7.08|
|test_wer | 15.35|15.72|

||TDNN-F(Pytorch, Adam, lda + dropout + ivector )|TDNN-F(Pytorch, Adam, lda dropout  without ivector ) |TDNN-F(Pytorch, Adam, dropout + ivector ) #3988  | TDNN-F(Pytorch, Adam, dropout  without ivector ) #3988 |
|--|-- | --|-- | --|
|dev_cer | 5.96|6.11|6.00 | 6.13|
|dev_wer | 13.74|13.86|13.72 | 13.97|
|test_cer |7.13 |7.21|7.20 | 7.19|
|test_wer |15.45 |15.49|15.52|15.55|
|global objf|-0.037806|-0.041149|-0.046589|-0.050291|
|valid objf|-0.038112(train_subset.scp)|-0.040516(train_subset.scp)|-0.053605|-0.062898|
|output-affine-norm|77.86|81.06|76.83|83.2|

@csukuangfj @qindazhu @danpovey please have a reivew.

### CODE REVIEW

**Comment by danpovey** [comment]
I like the get-raw-egs by chain2 part.
What I am not so crazy about is the part about supporting LDA.  This is a piece of complexity that we don't need.  We should implement delta features as part of the network.  I have previously explained how to do it... has it been done?  If not, perhaps @csukuangfj  could find the chat messages where I explained it and maybe show it to @fanlu or someone else?

**Comment by csukuangfj** [comment]
The implementation of delta-delta feature is in my ctc pullrequest.

I am currently porting the CTC CRF loss: http://oa.ee.tsinghua.edu.cn/~ouzhijian/pdf/ctc-crf.pdf

I find that it has a very good performance on the AIShell dataset, though the training
part is super slow. I am trying to make it run faster.

**Comment by fanlu** [comment]
OK, Let me try with add_deltas_layer

**Comment by fanlu** [comment]
||TDNN-F(Pytorch, Adam, lda + dropout + ivector )|TDNN-F(Pytorch, Adam, lda dropout  without ivector ) |TDNN-F(Pytorch, Adam, delta + dropout + ivector )|TDNN-F(Pytorch, Adam, delta dropout  without ivector ) |
|--|-- | --|-- | --|
|dev_cer | 5.95|6.12|5.95| 6.10|
|dev_wer | 13.73|13.91|13.67 | 13.86|
|test_cer |7.04 |7.22|7.18 | 7.14|
|test_wer |15.41 |15.62|15.60|15.49|

@danpovey @csukuangfj  please have a reivew

File: egs/aishell/s10/local/run_chain.sh [hunk 1]
```diff
@@ -191,33 +193,35 @@ if [ $stage -le 16 ]; then
...
+if [[ "$feat_type" == "lda" ]]; then
+  if [[ $stage -le 17 ]]; then
+    echo $dir
+    mkdir -p $dir/configs
+    if $train_ivector; then
+      cat <<EOF > $dir/configs/init.config
 input-node name=ivector dim=100
-input-node name=input dim=40
-output-node name=output input=Append(Offset(input, -1), input, Offset(input, 1), ReplaceIndex(ivector, t, 0))
+input-node name=input dim=120
```
Comment (csukuangfj) [inline egs/aishell/s10/local/run_chain.sh]: Is it possible to read 120 from feats.scp?
E.g., 3 x `feat-dim`

File: egs/aishell/s10/local/run_chain.sh [hunk 2]
```diff
@@ -123,20 +125,23 @@ if [[ $stage -le 12 ]]; then
...
 
-if [[ $stage -le 14 ]]; then
-  echo "creating denominator FST"
-  copy-transition-model $tree_dir/final.mdl $dir/0.trans_mdl
-  cp $tree_dir/tree $dir 
-  $train_cmd $dir/log/make_den_fst.log \
-    chain-make-den-fst $dir/tree $dir/0.trans_mdl $dir/phone_lm.fst \
-       $dir/den.fst $dir/normalization.fst || exit 1
+  echo "$0: creating phone language-model"
+  $train_cmd $dir/den_fsts/log/make_phone_lm_default.log \
```
Comment (qindazhu) [inline egs/aishell/s10/local/run_chain.sh]: I guess it's better if you can define `lang="default"` at the top of this file and replace `default` with `${lang}`

File: egs/aishell/s10/local/run_chain.sh [hunk 3]
```diff
@@ -159,57 +164,95 @@ log_level=info # valid values: debug, info, warning
...
-      ark,scp:$dir/merged_egs/cegs.JOB.ark,$dir/merged_egs/cegs.JOB.scp || exit 1
+      ark,scp:$dir/$merged_egs_dir/cegs.JOB.ark,$dir/$merged_egs_dir/cegs.JOB.scp || exit 1
 
-  rm $dir/egs/cegs.*.ark
+  rm $dir/raw_egs/cegs.*.ark
 fi
 
-feat_dim=$(cat $dir/egs/info/feat_dim)
+
+feat_dim=$(cat $dir/raw_egs/info.txt | grep 'feat_dim' | awk '{print $NF}')
```
Comment (qindazhu) [inline egs/aishell/s10/local/run_chain.sh]: Define `info_file=$dir/raw_egs/info.txt` and use it below may be better.

File: egs/aishell/s10/local/run_chain.sh (hunk 1 above)
Comment (fanlu) [reply egs/aishell/s10/local/run_chain.sh]: Done

File: egs/aishell/s10/local/run_chain.sh (hunk 2 above)
Comment (fanlu) [reply egs/aishell/s10/local/run_chain.sh]: Done

File: egs/aishell/s10/local/run_chain.sh (hunk 3 above)
Comment (fanlu) [reply egs/aishell/s10/local/run_chain.sh]: Done

**Comment by danpovey** [comment]
Great.. it looks like delta is as good as LDA.  Let's use delta features inside the network and not support LDA in these scripts, in order to keep it simple.

**Comment by fanlu** [comment]
@danpovey  done.

**Comment by danpovey** [comment]
Thanks!!  I am merging.  If anyone finds any issues with this we can fix them later.

**Comment by qindazhu** [comment]
Thanks @fanlu , I'll implement egs dataloader based on your script.

**Comment by fanlu** [comment]
> Thanks @fanlu , I'll implement egs dataloader based on your script.

Cool!!!
